### ISSUE TITLE
Successfully made mAP 75.8 as original yolo-v2

### ISSUE THREAD

**Issue by TPFRL** [marvis/pytorch-yolo2#151]
I know that this project is outdated, but I wanted to recover original mAP 75.8%, as succeed by Chainer. [https://github.com/chainer/chainercv/tree/master/examples/yolo](url)

If you change 'load weight' code just like as implementation of Chainer, you can get 75.8% when you use pretrained weight in [https://pjreddie.com/darknet/yolov2/](url)
![image](https://user-images.githubusercontent.com/8455454/54654811-ce9e2800-4b02-11e9-822b-e83d49429516.png)
(You have to change the _devkit_path in voc_eval.py to use given [Evaluate the model](https://github.com/marvis/pytorch-yolo2#evaluate-the-model) )https://github.com/marvis/pytorch-yolo2/blob/6c7c1561b42804f4d50d34e0df220c913711f064/scripts/voc_eval.py#L205

Change darknet.py
```python
    def load_weights(self, weightfile):
        with open(weightfile, mode='rb') as f:
            major = np.fromfile(f, dtype=np.int32, count=1)
            minor = np.fromfile(f, dtype=np.int32, count=1)
            np.fromfile(f, dtype=np.int32, count=1)  # revision
            if major * 10 + minor >= 2 and major < 1000 and minor < 1000:
                np.fromfile(f, dtype=np.int64, count=1)  # seen
            else:
                np.fromfile(f, dtype=np.int32, count=1)  # seen

            ind = -2
            for block in self.blocks:
                if ind >= len(self.models):
                    break
                ind = ind + 1
                if block['type'] == 'net':
                    continue
                elif block['type'] == 'convolutional':
                    model = self.models[ind]
                    batch_normalize = int(block['batch_normalize'])
                    if batch_normalize:
                        start = load_conv_bn(f, model[0], model[1])
                    else:
                        start = load_conv(f, model[0])
                elif block['type'] == 'connected':
                    model = self.models[ind]
                    if block['activation'] != 'linear':
                        start = load_fc(f, model[0])
                    else:
                        start = load_fc(f, model)
                elif block['type'] == 'maxpool':
                    pass
                elif block['type'] == 'reorg':
                    pass
                elif block['type'] == 'route':
                    pass
                elif block['type'] == 'shortcut':
                    pass
                elif block['type'] == 'region':
                    pass
                elif block['type'] == 'avgpool':
                    pass
                elif block['type'] == 'softmax':
                    pass
                elif block['type'] == 'cost':
                    pass
                else:
                    print('unknown type %s' % (block['type']))
```

And change cfg.py
```python
def load_param(file, param):
    param.data.copy_(torch.from_numpy(
        np.fromfile(file, dtype=np.float32, count=param.numel()).reshape(param.shape)
    ))

def load_conv(file, conv_model):
    load_param(file, conv_model.bias)
    load_param(file, conv_model.weight)

def load_conv_bn(file, conv_model, bn_model):
    load_param(file, bn_model.bias)
    load_param(file, bn_model.weight)
    load_param(file, bn_model.running_mean)
    load_param(file, bn_model.running_var)
    load_param(file, conv_model.weight)
```

**Comment by abhigoku10** [comment]
@TPFRL  HI just wanted ot know how u achieved this mAP did u do any augmentation and how much was your training data and validation data

**Comment by TPFRL** [comment]
@abhigoku10 I used darknet's pretrained weight.
The previous version of this repository couldn't reproduce darknet's mAP, even if you use provided pretrained weight.

**Comment by puyangma** [comment]
@TPFRL Thank you for sharing this. May I ask which pytorch and python version are you using for running these updates? There are some issues when I tried pytorch 0.3.1 and python 2.7

**Comment by TPFRL** [comment]
@puyangma Oh, I didn't write it...
I used pytorch 1.0 and python 3. I needed to change some parts to run on this environment.

I wrote like this because I thought other versions of torch and python can also run this code..
If there is errors from other parts from this, you should change the codes.

**Comment by puyangma** [comment]
@TPFRL Thanks for your reply! I wonder if you were able to run the training part with pytorch 1.0 and python 3? If so, would you mind sharing what changed you made to the source code? I have been trying without success to debug the training part with pytorch 0.4...

**Comment by TPFRL** [comment]
@puyangma Was able to... But not perfectly yet. Could increase mAP in VOC dataset, but couldn't in Imagenet.

I was able to make this repository's code to be compatible with torch1.0 and python3,
-Make sure you change int division '/' to '//'
-Make sure you correct the dimensions. I don't know how these worked in previous versions, but there are dimension differences. For example, you have to change region_loss.py 142~145's '.data' to '.flatten()' and 152's 'data[0]' to 'data', etc... 
I don't remember why I changed region_loss.py 126~131 like below, but maybe those were also some dimension difference issue. Anyway, below is more straightforward :
```python
        output   = output.view(nB, nA, (5+nC), nH, nW)
        x    = F.sigmoid(output[:, :, 0, ])
        y    = F.sigmoid(output[:, :, 1, ])
        w    = output[:, :, 2, ]
        h    = output[:, :, 3, ]
        conf = F.sigmoid(output[:, :, 4, ])
        indicesofcls = torch.linspace(5,5+nC-1,nC).long() #must be long or byte
        cls = output[:, :, indicesofcls, ]
```

Actually, I compared this code with darknet's original C code, and found there were some difference in logic. So I changed the logic, and I don't remember which part is the part I changed for compatibility, and which part is for changing logic...
For example, the easiest difference is that this repository's jitter is "fixed" to 0.2
And It doesn't treat the "seen<12800" case perfectly. In region_loss.py 48 it is filled with 1 but actually I think it should be filled with 0.01.

**Comment by puyangma** [comment]
@TPFRL Thanks a lot! Yes your modifications to the eval part works pretty well! If you don't mind, I have a final question about training. My total loss during training does not decrease to 0.xxx, so I wonder about how many epochs did your training logs show 0.xxx total loss? Did you change any of the following parameters in the yolo-voc.cfg before calling train?

batch_size=64
subdivision =8
steps=-1,500,40000,60000
scales=0.1,10,.1,.1

**Comment by TPFRL** [comment]
@puyangma Um... for me, total loss didn't go down to that much.
I think even in original darknet you cannot make loss low that much...

**Comment by puyangma** [comment]
@TPFRL Was your lowest total loss around 20? Mine gets stuck around 50.

**Comment by TPFRL** [comment]
@puyangma sometimes it went down around 10~20, sometimes it was around 50... it fluctuated.
And I changed the loss function slightly, so maybe mine isn't a good criteria

**Comment by datbu178** [comment]
@TPFRL  Thank you for sharing...Can you tell me the mAP before you change the code? And how many epochs did you train?
I'm just get mAP 61.6% after 100 epochs (I haven't changed the code yet)...

**Comment by TPFRL** [comment]
@datbu178 um.. actually I haven't tried training from scratch.

**Comment by puyangma** [comment]
@datbu178 May I ask what is your GPU memory? What is the batch size you used to train for more than 100 epochs? My GPU is Tesla K80 which has 24GB but I still encountered memory out of error after training for 60 epochs. I already decreased the batch size from 64 to 32.

**Comment by datbu178** [comment]
@puyangma yes, it will out of memory when the size of the input image is changed ... So I changed the code in dataset.py (from line 38 to line 53) so that the input image size is always 13 * 32 ...
Maybe that's why my mAP is so low ???

**Comment by puyangma** [comment]
@datbu178 Are you using the VOC dataset? Shouldn't the input image size be 416 by 416? You mean it will generate OOM due to multi-scale training?

**Comment by datbu178** [comment]
@puyangma Yes...Please check the code in dataset.py....when the input image size is increased, OOM will happen.

**Comment by puyangma** [comment]
@datbu178 I was training with batch size 32, which is already half of that used in the repo. https://github.com/marvis/pytorch-yolo2/issues/85 
Someone used a 11GB GPU and successfully trained with batch size 32.

**Comment by puyangma** [comment]
@datbu178 It seems from some posts that using batch size = 16 generates mAP around 62 or so.  What was your batch size for your reported mAP 61.6%?

**Comment by datbu178** [comment]
@puyangma My batch size is 25 and I train with the fixed input image size  416x416 ....  And I achieved mAP 61.6% after 100 epochs. 
I'm trying to train in multi-scale mode, the result will be updated later.

**Comment by puyangma** [comment]
@datbu178 Thank you. Yes I think the multi-scale training and the relatively small batch size are the reason for lower mAP.

**Comment by datbu178** [comment]
Update: I train in multi-scale mode with batch size 16...After 230 epochs, my mAP is 72.9%

**Comment by puyangma** [comment]
@datbu178 Wonderful, congrats! When you train in multi-scale mode, did you change dataset.py at all or did you just use it as it is? Except for batch size changed to 16, did you change any of these following parameters?

learning_rate=0.001
burn_in=1000
max_batches = 80200
policy=steps
steps=-1,500,40000,60000
scales=0.1,10,.1,.1

**Comment by datbu178** [comment]
@puyangma Thank you, I changed the parameters as follows:

learning_rate = 0.0003
burn_in=1000
max_batches = 80200
policy=steps
steps=-1,1500,120000,180000
scales=0.1,10,.1,.1

I did not change the other parameters. 
I think I will try to find out the reasons why it did not achieve the mAP 75.8%. For example, due to smaller batch sizes,....

**Comment by puyangma** [comment]
@datbu178 Perhaps using earlier weights like 200 or 190 could achieve higher mAP due to overfit. Are you using Python 3.0 and Pytorch 4.0 to run training?

**Comment by datbu178** [comment]
@puyangma Python 3.0 
Pytorch 0.4.1

**Comment by puyangma** [comment]
@datbu178 oh right I meant to say pytorch 0.4. I have tried but failed to make the original training part be compatible with Pytorch 0.4.1 and Python 3.0. Using Pythorch 0.3 seems to give me lower mAP. Would you mind sharing with me all your training codes? So that I can try different warm-up or learning rate and see if mAP 75.8% can be achieved. My email is ma.puyang@gmail.com

**Comment by datbu178** [comment]
@puyangma Sure... but I think it's an algorithmic problem ... The algorithm in this project may not be exactly the same as the original yolov2 project.

**Comment by xingyueye** [comment]
Beautiful job, could you upload your yolo.weights? thans

**Comment by xingyueye** [comment]
@TPFRL great job, could you upload your yolo.weights with mAP 75? thanks

**Comment by TPFRL** [comment]
@datbu178 Exactly. The algorithm is a little different.
@xingyueye You can find them in official website, such as [https://pjreddie.com/darknet/yolov2/](url)

**Comment by xingyueye** [comment]
@TPFRL ok，thanks. Do you know why can't this project get the same mAP 76.8 like darknet.

**Comment by TPFRL** [comment]
@xingyueye Not sure but you can check here. [https://github.com/chainer/chainercv/tree/master/examples/yolo](https://github.com/chainer/chainercv/tree/master/examples/yolo)
